File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/4sum/code/python/4sum.py
# ============================================================
# 4Sum
# ============================================================

from typing import List

# ============================================================
# Solution 1: Brute Force — All Quadruplets
# Time: O(N⁴) | Space: O(1)
# ============================================================
class Solution1:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        result = set()
        n = len(nums)
        for i in range(n-3):
            for j in range(i+1,n-2):
                for k in range(j+1,n-1):
                    for l in range(k+1,n):
                        if nums[i]+nums[j]+nums[k]+nums[l]==target:
                            result.add((nums[i],nums[j],nums[k],nums[l]))
        return [list(t) for t in result]

# ============================================================
# Solution 2: Sort + Two Nested Loops + Two Pointers (Optimal)
# Time: O(N³) | Space: O(1) extra
# ============================================================
class Solution2:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        result = []
        n = len(nums)

        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i-1]: continue

            for j in range(i+1, n-2):
                if j > i+1 and nums[j] == nums[j-1]: continue

                left, right = j+1, n-1
                while left < right:
                    total = nums[i]+nums[j]+nums[left]+nums[right]

                    if total == target:
                        result.append([nums[i],nums[j],nums[left],nums[right]])
                        while left<right and nums[left]==nums[left+1]: left+=1
                        while left<right and nums[right]==nums[right-1]: right-=1
                        left+=1; right-=1
                    elif total < target: left+=1
                    else: right-=1

        return result

File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/4sum/code/cpp/4sum.cpp
// ============================================================
// 4Sum
// ============================================================

#include <vector>
#include <algorithm>
using namespace std;

// ============================================================
// Solution 1: Brute Force — All Quadruplets
// Time: O(N⁴) | Space: O(1)
// ============================================================
class Solution1 {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        set<vector<int>> result;
        int n = nums.size();

        for (int i = 0; i < n-3; i++)
            for (int j = i+1; j < n-2; j++)
                for (int k = j+1; k < n-1; k++)
                    for (int l = k+1; l < n; l++)
                        if ((long)nums[i]+nums[j]+nums[k]+nums[l] == target)
                            result.insert({nums[i],nums[j],nums[k],nums[l]});

        return vector<vector<int>>(result.begin(), result.end());
    }
};

// ============================================================
// Solution 2: Sort + Two Nested Loops + Two Pointers (Optimal)
// Time: O(N³) | Space: O(1) extra
// ============================================================
class Solution2 {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        int n = nums.size();

        for (int i = 0; i < n - 3; i++) {
            // Skip duplicate i
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < n - 2; j++) {
                // Skip duplicate j
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                int left = j + 1, right = n - 1;

                while (left < right) {
                    // Use long to prevent integer overflow
                    long sum = (long)nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        result.push_back({nums[i], nums[j], nums[left], nums[right]});

                        // Skip duplicates
                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }
};
